clients/src/main/java/org/apache/kafka/common/requests/MetadataResponse.java

```diff
@@ -144,63 +132,155 @@ public MetadataResponse(Struct struct) {
+            } else if (metadata.error == Errors.TOPIC_AUTHORIZATION_FAILED) {
+                unauthorizedTopics.add(metadata.topic);
+            }
         }
-        return unauthorizedTopics;
+
+        return new Cluster(this.brokers, partitions, unauthorizedTopics);
     }
 
-    public Map<String, Errors> errors() {
-        return this.errors;
+    /**
+     * Get all brokers returned in metadata response
+     * @return the brokers
+     */
+    public Collection<Node> brokers() {
+        return brokers;
     }
 
-    public boolean hasError(String topic) {
-        return this.errors.containsKey(topic);
+    public static MetadataResponse parse(ByteBuffer buffer) {
+        return new MetadataResponse(CURRENT_SCHEMA.read(buffer));
     }
 
-    public Cluster cluster() {
-        return this.cluster;
+    public static class TopicMetadata {
+        private final Errors error;
+        private final String topic;
+        private final List<PartitionMetadata> partitionMetadata;
+
+        public TopicMetadata(Errors error,
+                             String topic,
+                             List<PartitionMetadata> partitionMetadata) {
+            this.error = error;
+            this.topic = topic;
+            this.partitionMetadata = partitionMetadata;
+        }
+
+        public Errors error() {
+            return error;
+        }
+
+        public String topic() {
+            return topic;
+        }
+
+        public List<PartitionMetadata> partitionMetadata() {
+            return partitionMetadata;
+        }
+
     }
 
-    public static MetadataResponse parse(ByteBuffer buffer) {
-        return new MetadataResponse(CURRENT_SCHEMA.read(buffer));
+    public static class PartitionMetadata {
+        private final Errors error;
+        private final int partition;
+        private final Node leader;
+        private final List<Node> replicas;
+        private final List<Node> isr;
+
+        public PartitionMetadata(Errors error,
+                                 int partition,
+                                 Node leader,
+                                 List<Node> replicas,
+                                 List<Node> isr) {
+            this.error = error;
+            this.partition = partition;
+            this.leader = leader;
+            this.replicas = replicas;
+            this.isr = isr;
+        }
+
+        public Errors error() {
+            return error;
+        }
+
+        public int partition() {
+            return partition;
+        }
+
+        public Node leader() {
+            return leader;
+        }
+
+        public List<Node> replicas() {
+            return replicas;
+        }
+
+        public List<Node> isr() {
+            return isr;
+        }
+
+    }
+
+    private static Node[] toArray(Collection<Node> nodes) {
+        return nodes.toArray(new Node[nodes.size()]);
```

Comment:
So, it surprisingly turns out that allocating the array with the right size is actually slower than passing an empty array to this method:

http://shipilev.net/blog/2016/arrays-wisdom-ancients/